### PR TITLE
[Merged by Bors] - feat(RingTheory/Etale): etale <=> separable for field extensions of finite type.

### PR DESCRIPTION
**Pull request by erdOne**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/RingTheory/Etale/Field.lean [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+# Etale algebras over fields
+
+## Main results
+
+Let `K` be a field, `A` be a `K`-algebra and `L` be a field extension of `K`.
+
+- `Algebra.FormallyEtale.of_isSeparable`:
+    If `L` is separable over `K`, then `L` is formally etale over `K`.
+- `Algebra.FormallyUnramified.iff_isSeparable`:
+    If `L` is (essentially) of finite type over `K`, then `L/K` is etale iff `L/K` is separable.
```
Comment (jcommelin) [inline Mathlib/RingTheory/Etale/Field.lean:19]:
```suggestion
- `Algebra.FormallyEtale.iff_isSeparable`:
    If `L` is (essentially) of finite type over `K`, then `L/K` is etale iff `L/K` is separable.
```

File: Mathlib/RingTheory/Etale/Field.lean [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    obtain ⟨α, hx, hy⟩ := H x y
+    simp only [← hg₃ _ _ hx, ← hg₃ _ _ hy, ← map_add, ← hg₃ _ _ (add_mem hx hy)]
+    rfl
+  · intro r
+    show g _ (algebraMap K _ r) = _
+    rw [AlgHom.commutes]
+  · ext x
+    simpa using AlgHom.congr_fun (hg₁ x) (IntermediateField.AdjoinSimple.gen K x)
+
+theorem iff_isSeparable [EssFiniteType K L] :
```
Comment (jcommelin) [inline Mathlib/RingTheory/Etale/Field.lean:129]: Would it make sense to rephrase this as a `tfae` including `FormallyEtale`, `FormallyUnramified` and `IsSeparable`?

Comment (erdOne) [reply Mathlib/RingTheory/Etale/Field.lean:129]: I don't think `tfae` lemmas are that useful application wise. But it might be a good sanity check and I'll add it once we have the equivalence with smooth-ness.

File: Mathlib/RingTheory/Etale/Field.lean [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+variable (K A L : Type u) [Field K] [Field L] [CommRing A] [Algebra K A] [Algebra K L]
+
+open Algebra Polynomial
+
+open scoped TensorProduct
+
+namespace Algebra.FormallyEtale
+
+theorem of_isSeparable_aux [Algebra.IsSeparable K L] [EssFiniteType K L] :
```
Comment (kbuzzard) [inline Mathlib/RingTheory/Etale/Field.lean:38]: Maybe this should have a docstring or a comment saying "don't use this, it's a weaker version of `of_isSeparable`?

File: Mathlib/RingTheory/Etale/Field.lean [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+    use u
+    rw [← Ideal.Quotient.eq_zero_iff_mem, map_add, map_mul, map_one, hu, mul_neg,
+      IsUnit.mul_val_inv, neg_add_cancel]
+  use pb.liftEquiv.symm ⟨x + u * aeval x (minpoly K pb.gen), ?_⟩
+  · apply pb.algHom_ext
+    simp [hx, hx']
+  · rw [← eval_map_algebraMap, Polynomial.eval_add_of_sq_eq_zero, derivative_map,
+      ← one_mul (eval x _), eval_map_algebraMap, eval_map_algebraMap, ← mul_assoc, ← add_mul,
+      ← Ideal.mem_bot, ← h, pow_two, add_comm]
+    exact Ideal.mul_mem_mul hu (Ideal.Quotient.eq_zero_iff_mem.mp hx')
```
Comment (kbuzzard) [inline Mathlib/RingTheory/Etale/Field.lean:65]:
```suggestion
    · exact Ideal.mul_mem_mul hu (Ideal.Quotient.eq_zero_iff_mem.mp hx')
```
I thought we had a linter which would moan about this sort of thing now? But it didn't so I'll moan about it :-)

**Review by kbuzzard — COMMENTED** [review]
I guess I'd be happier if these long proofs (one of them is > 50 lines) had some kind of comments in, or at least a *precise* reference to Iversen rather than just a reference to the entire book. But other than that this is fine.

File: Mathlib/RingTheory/Etale/Field.lean (hunk 3 above)
Comment (erdOne) [inline Mathlib/RingTheory/Etale/Field.lean:38]:
```suggestion
/--
This is a weaker version of `of_isSeparable` that additionally assumes `EssFiniteType K L`.
Use that instead.
-/
theorem of_isSeparable_aux [Algebra.IsSeparable K L] [EssFiniteType K L] :
```

**Comment by erdOne** [comment]
Okay I added some comments and a reference.

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge
